org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/Document.xtend

```diff
@@ -119,31 +119,36 @@ import org.eclipse.lsp4j.Range
     }
 
     def String getSubstring(Range range) {
-    		val start = getOffSet(range.start)
-    		val end = getOffSet(range.end)
-    		return this.contents.substring(start, end)
+          val start = getOffSet(range.start)
+          val end = getOffSet(range.end)
+          return this.contents.substring(start, end)
     }
 
     def Document applyChanges(Iterable<? extends TextEdit> changes) {
-        var newContent = contents
+        var currentDocument = this
         for (change : changes) {
-            if (change.range === null) {
-                newContent = change.newText
-            } else {
-                val start = getOffSet(change.range.start)
-                val end = getOffSet(change.range.end)
-                newContent = newContent.substring(0, start) + change.newText + newContent.substring(end)
-            }
+            val newContent = if (change.range === null) {
+                    change.newText
+                } else {
+                    val start = currentDocument.getOffSet(change.range.start)
+                    val end = currentDocument.getOffSet(change.range.end)
+                    currentDocument.contents.substring(0, start) + change.newText + currentDocument.contents.substring(end)
+                }
+            val newVersion = if (currentDocument.version !== null)
+                    currentDocument.version + 1
+                else
+                    null
+            currentDocument = new Document(newVersion, newContent)
```

Comment:
Here we are forgetting the value of printSourceOnError

Reply:
Also it appears the we should create only a single new version (see above)

Reply:
Fixed